accepted/00000-modular-repos-with-clm.md

```diff
@@ -0,0 +1,180 @@
+
+For each key of the map, the backend gets the packages with the same name from all available sources. If any of the packages are not included in the corresponding artifacts list, the package is added to the blacklist.
+
+As a result, when a module defines a package name in its API *and* provides an actual RPM with it, all the other RPMs with the same name are effectively blacklisted even if they are in a different source.
+
+### Module updates and multiple versions
+Every released update of a module is added into the module metadata as a separate module entry with a new `version` number. By design, Uyuni needs to include either all versions of a module or none at all.
+This can be achieved by enabling all versions of the module in the API at the same time. As a result, all the artifacts from all versions will be included sharing the same key in the returning map.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+For each build of the project, the backend needs a full list of RPMs that each module provides. The most important drawback is the amount of data to process for each build considering the module metadata will grow constantly with each released update.
+
+# Alternatives
+[alternatives]: #alternatives
+
+The current version of Uyuni supports modular repositories in the most basic way. Modular repositories are recognized, mirrored correctly and can be assigned systems. However, doing package operations on these systems through the Uyuni UI is not supported. These systems must be managed from their own package managers.
+
+The next step to supporting modular repositories would be to implement modularity in the core with dedicated logic for package operations and a new set of UI. The most obvious benefit of this would be to be able to do package operations through the UI while still providing modular repositories to the clients. However, this implementation would require much more effort and currently beyond the scope.
+
+# Further enhancements
+[enhancements]: #enhancements
+
+### Interactive inputs for filters
+To provide better user experience, the filter inputs can be turned into responsive dropdowns showing a list of module/stream names with realtime feedback.
+However, since the module list can only be retrieved from the project sources and the current implementation of CLM filters are independent of the sources, this is not currently achievable.
+
+### Selective version filtering
+If required, it is possible to selectively filter a specific version of any module by introducing additional types of module filters.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+ - How to communicate between the Python API and the Java backend
+ - What should be the behavior for the unselected modules? Should the modules be disabled unless there is a filter targeting them? Should they be enabled with default streams in the best effort?
```

Comment:
What does Fedora/RHEL/CentOS do? We should do the same

Reply:
dnf automatically figures out and enables the default stream for the module when a package install is requested. Since SUMA does not recognize modularity at the package operations level, this is not possible for us.

The question is about UX. The point is to set all the modules either enabled or disabled. Let's say I have 10 modules available and I added 4 filters to select streams for 4 modules. What happens to the remaining 6?
 - Should we disable them?
 - Should we enable them with default streams? What if this combination is not possible?

Reply:
My suggestion: disabled by default.

I think it is always possible to disable everything, but not to enable everything.